engine/immediate_task_concurrent_processor.go

```diff
@@ -66,17 +67,48 @@ func (w *immediateTaskConcurrentProcessor) GetTasksToProcessChan() chan<- data_m
 func (w *immediateTaskConcurrentProcessor) AddImmediateTaskQueue(
 	shardId int32, tasksToCommitChan chan<- data_models.ImmediateTask,
 ) (alreadyExisted bool) {
-	exists := w.currentShards[shardId]
-	w.currentShards[shardId] = true
-	w.taskToCommitChans[shardId] = tasksToCommitChan
+	lock := sync.RWMutex{}
```

Comment:
the lock needs to be inited in constructor and kept as a member field. see this simple example: https://github.com/uber/cadence/blob/02c7efbed448c4a493b3a971e8e0e292e17c6d91/common/collection/concurrentPriorityQueue.go#L27

Reply:
But what's the issue with the current implementation? It seems to me create a lock when needed is also doable. I am wondering if this is mainly about performance?